src/Microsoft.OData.Core/Json/JsonLightInstanceAnnotationWriter.cs

```diff
@@ -84,27 +84,97 @@ internal JsonLightInstanceAnnotationWriter(ODataJsonLightValueSerializer valueSe
 
-            HashSet<string> instanceAnnotationNames = new HashSet<string>(StringComparer.Ordinal);
-            foreach (var annotation in instanceAnnotations)
+            // this method runs in a hot path hence the optimizations
+
+            // initialize to null to avoid instantiating a HashSet if there are no annotations
+            HashSet<string> instanceAnnotationNames;
+            if (instanceAnnotations?.Count > 0)
+            {
+                instanceAnnotationNames = new HashSet<string>(StringComparer.Ordinal);
+            }
+            else
+            {
+                return;
+            }
+
+            // this method is called with a List most of the time
+            // foreach against a List does not allocate the enumerator to the heap,
+            // but foreach against an IEnumerable does
+            if (instanceAnnotations is List<ODataInstanceAnnotation> instanceAnnotationsList)
+            {
+                foreach (ODataInstanceAnnotation annotation in instanceAnnotationsList)
+                {
+                    this.WriteAndTrackInstanceAnnotation(annotation, tracker, instanceAnnotationNames, ignoreFilter, propertyName);
+                }
+            }
+            else
             {
-                if (!instanceAnnotationNames.Add(annotation.Name))
+                foreach (ODataInstanceAnnotation annotation in instanceAnnotations)
                 {
-                    throw new ODataException(ODataErrorStrings.JsonLightInstanceAnnotationWriter_DuplicateAnnotationNameInCollection(annotation.Name));
+                    this.WriteAndTrackInstanceAnnotation(annotation, tracker, instanceAnnotationNames, ignoreFilter, propertyName);
                 }
+            }
+        }
+
+        /// <summary>
+        /// Write instance annotation if not already written.
+        /// </summary>
+        /// <param name="annotation">The instance annotation to write.</param>
+        /// <param name="tracker">The tracker to track if instance annotations are written.</param>
+        /// <param name="instanceAnnotationNames">Set used to detect a duplicate annotation.</param>
+        /// <param name="ignoreFilter">Whether to ignore the filter in settings.</param>
+        /// <param name="propertyName">The name of the property this instance annotation applies to.</param>
+        private void WriteAndTrackInstanceAnnotation(
+            ODataInstanceAnnotation annotation,
+            InstanceAnnotationWriteTracker tracker,
+            HashSet<string> instanceAnnotationNames,
+            bool ignoreFilter = false,
+            string propertyName = null)
+        {
 
-                if (!tracker.IsAnnotationWritten(annotation.Name)
-                    && (!ODataAnnotationNames.IsODataAnnotationName(annotation.Name) || ODataAnnotationNames.IsUnknownODataAnnotationName(annotation.Name)))
+            if (!instanceAnnotationNames.Add(annotation.Name))

```

Comment:
Because we also want to add the annotation to the set. If adding returns false, it means the item was not added and duplicate was found. If it returns true, it means it was added successfully. So this saves us from having separate calls to `Contains` and `Add`.